### PR TITLE
Add a new Image class which allows using images in GLSL

### PR DESCRIPTION
**Pull request by robquill**
Create a Memory base-class which Image and Tensor inherit from.

Operations and Sequences now use the Memory base-class instead of the Tensor class.

OpTensorCopy has been renamed to OpCopy and supports copying between Tensors and Tensors, Tensors and Images, Images and Tensors and Images and Images.

OpTensorSyncDevice/Local has been renamed to OpSyncDevice/Local and supports synchronising both Tensors and Images.

### CODE REVIEW

**Comment by robquill** [comment]
General comments (from @axsaucedo)
 
Renaming of classes Tensor->Vec, Image->Mat:
See below for context on thinking
Let's discuss if doing renaming now or after merge 
Namely as reviewing renaming PR easier than together with all changes
 
height, weight -> x,y.
Would there also be an Mat3D impl (in the future)?
 
Memory object:
Parent class impl makes sense - naming convention, is Memory the best representation?
Keen to explore this question as it will become core to framework
What are alternatives considered?
Why does Memory.hpp have #include <memory>?
Remember to prefix kp:: all Kompute objects even inside namespace (incl other files)
We've been looking as an opportunity to expose the underlying buffer (so people can bring their own)
However this to be approached as separate initiative / PR
 
Manager.hpp:
Why different map to manage images vs vectors vs single mem one?
 
Tests:
Why on tests changing TensorT->Memory? ie lose functionality from abstraction
Let's run codecov to ensure we can look to have good coverage
Test custom dtype for vectors vs image?
 
Operations:
Currently requires explicitly specifying OpASync[...] or OpACopyB
Is there a way to implement an OpCopy or OpSync that don't require explicitly selecting?
I haven't properly thought about it being possible / not possible but wanted to ask as adds cognitive load on api
(Relevant to question above of why changing Tensor->Memory on tests)
 
On the point for "A more recent issue asks for 8-bit data support":
Currently adding an extra datatype should be pretty straightforward
See https://github.com/KomputeProject/kompute/blob/a1555854bb035feb9f15c3049224d2f6dd961409/test/TestOpShadersFromStringAndFile.cpp#L12-L59
Is there something you meant further to just adding uint8_t to the supported datatypes?
@SimLeek, on this point "Casting everything to float through python and back is basically a hack":
I'm trying to understand what you mean, from the current impl. it would be using numpy dtypes, do you mean when converting to array?
Also on this point "Oh, also, I kinda want custom data types as a different issue, like sending structs to buffers.":
Does the link above answer the question or you meant something else?

**Comment by robquill** [comment]
TODO:

- [ ] Update examples (they don't build)
- [ ] height, weight -> x,y.
- [ ] Why does Memory.hpp have #include ?
- [ ] Remember to prefix kp:: all Kompute objects even inside namespace (incl other files)
- [ ] Manager.hpp: Why different map to manage images vs vectors vs single mem one?
- [ ] Let's run codecov to ensure we can look to have good coverage
- [ ] Test custom dtype for vectors vs image?
- [ ] Currently requires explicitly specifying OpASync[...] or OpACopyB. Is there a way to implement an OpCopy or OpSync that don't require explicitly selecting?
- [ ] Update / extend the ref diagram (bottom right) as progresses: https://app.diagrams.net/#G1mRJlFEWpYD2dMU7j1HGwwDrxPeEmUjl8#%7B%22pageId%22%3A%22fwopQ7pWVPHFaTMxwvhy%22%7D

Answered:

- [ ] Why on tests changing TensorT->Memory? ie lose functionality from abstraction

The reason is that now the Sequence->eval uses Memory instead of Tensor and it looks like the compiler doesn't know how to convert std::vector<std::shared_ptr<Tensor>> to std::vector<std::shared_ptr<Memory>>. For example, this snippet from TestASyncOperations.cpp does not work:

```c++
    std::vector<std::shared_ptr<kp::Tensor>> inputsSyncB;
    ...
    sq->eval<kp::OpTensorSyncDevice>(inputsSyncB);
```
because:

```
/home/rob/dev/git/kompute/test/TestAsyncOperations.cpp:61:38: note:   cannot convert ‘inputsSyncB’ (type ‘std::vector<std::shared_ptr<kp::Tensor> >’) to type ‘std::vector<std::shared_ptr<kp::Memory> >’
```

Deferred:

- [ ] Renaming of classes Tensor->Vec, Image->Mat:
- [ ] naming convention, is Memory the best representation?
- [ ] We've been looking as an opportunity to expose the underlying buffer (so people can bring their own). However this to be approached as separate initiative / PR.

**Comment by axsaucedo** [comment]
> ### Test custom dtype for vectors vs image?
> 
> I'm not completely sure what this meant. Custom types don't make sense for images as you can only create them with the formats Vulkan/the HW supports. You can't have images that contain structs etc for example.

Good point, it seems indeed the Image type would not support the dtype, may be worth covering this as a compile-type error if anyone tries

**Review by axsaucedo — CHANGES_REQUESTED** [review]
Thank you for the PR @robquill - I've done an initial pass and added a few comments, let me know if any thoughts or questions

A few more points to add to the actions:

* [ ] Adding further python tests to identify issues/limits with tensor/image interactions
* [ ] Explore/discuss exposing further functionality through top level parent class
* [ ] Implement or extend 1-2 of the more advanced examples with image class to further battle test the limitations (can be done as a separate PR)
* [ ] Exploring how the height,width->x,y could also be exposed top level class (Tensor class could also have the same as default - i.e. x,y->x,1)

I've also been thinking about the design and I do feel like exposing more functionality to the top level class makes sense so it can be used without losing any functionality (+ simplifying API), so it's possible to have full abstraction on the underlying objects. Thinking that this could also allow for the python interface to enable for similar top level abstraction with this base class.

File: test/TestImage.cpp
Comment (robquill) [reply test/TestImage.cpp:18]: 3-channel images are fairly uncommon I would have thought, so we could just say they are not supported if you prefer. If there is an easy way to skip a test on a platform then if you let me know how I can do that.

File: src/Image.cpp
Comment (robquill) [reply src/Image.cpp:373]: I'm not sure. I just copied what was already there for Tensor. See https://github.com/KomputeProject/kompute/blob/40f9adf62e270d679ce5c9a72b0fae01155a094d/src/Tensor.cpp#L305.

File: src/include/kompute/Image.hpp
Comment (robquill) [reply src/include/kompute/Image.hpp:109]: I've added this to the list of deferred items in https://github.com/KomputeProject/kompute/pull/388#issuecomment-2291136436

File: python/test/test_image_types.py
Comment (robquill) [reply python/test/test_image_types.py]: Fixed.

**Comment by robquill** [comment]
> > ### Test custom dtype for vectors vs image?
> > I'm not completely sure what this meant. Custom types don't make sense for images as you can only create them with the formats Vulkan/the HW supports. You can't have images that contain structs etc for example.
> 
> Good point, it seems indeed the Image type would not support the dtype, may be worth covering this as a compile-type error if anyone tries

I'm not sure how to make this a compile-time error because the dtype is passed as a variable.

File: src/include/kompute/Image.hpp
Comment (robquill) [reply src/include/kompute/Image.hpp]: Fixed.

File: src/include/kompute/Image.hpp
Comment (robquill) [reply src/include/kompute/Image.hpp:61]: I didn't do this because there is a version below which doesn't take a tiling parameter. It infers what the tiling should be based on the memory type (host-accessible memory will be linear tiled, device-only memory will be optimal tiled).

**Comment by axsaucedo** [comment]
> I'm not sure how to make this a compile-time error because the dtype is passed as a variable.

Apologies, it seems I didn't share full context. In the current implementation there's nothing that would have to be done. However following the [suggestion here](https://github.com/KomputeProject/kompute/pull/388#discussion_r1720949108) of merging the imageType and tensorType into a top level enum it would be required as it would by default have an `eCustom` enum value (as this is possible for the Tensor class), so for the image it would be required to restrict it, which could be done by using something like [std::is_base_of](https://en.cppreference.com/w/cpp/types/is_base_of). 

However indeed it would only be necessary given the suggestion above - although hasn't yet been discussed. Let me know if this is not clear and I can expand.

**Comment by robquill** [comment]
> > I'm not sure how to make this a compile-time error because the dtype is passed as a variable.
> 
> Apologies, it seems I didn't share full context. In the current implementation there's nothing that would have to be done. However following the [suggestion here](https://github.com/KomputeProject/kompute/pull/388#discussion_r1720949108) of merging the imageType and tensorType into a top level enum it would be required as it would by default have an `eCustom` enum value (as this is possible for the Tensor class), so for the image it would be required to restrict it, which could be done by using something like [std::is_base_of](https://en.cppreference.com/w/cpp/types/is_base_of).
> 
> However indeed it would only be necessary given the suggestion above - although hasn't yet been discussed. Let me know if this is not clear and I can expand.

OK, let me see how much I can refactor into the Memory class then we can see what to do about this. I don't completely understand the purpose of merging Image and Tensor and having an enum to differentiate. You would just end up replacing dynamic_cast with a test of the enum, which I imagine is what the compiler ends up doing internally to do the dynamic_cast.

**Comment by axsaucedo** [comment]
> > > I'm not sure how to make this a compile-time error because the dtype is passed as a variable.
> > 
> > 
> > Apologies, it seems I didn't share full context. In the current implementation there's nothing that would have to be done. However following the [suggestion here](https://github.com/KomputeProject/kompute/pull/388#discussion_r1720949108) of merging the imageType and tensorType into a top level enum it would be required as it would by default have an `eCustom` enum value (as this is possible for the Tensor class), so for the image it would be required to restrict it, which could be done by using something like [std::is_base_of](https://en.cppreference.com/w/cpp/types/is_base_of).
> > However indeed it would only be necessary given the suggestion above - although hasn't yet been discussed. Let me know if this is not clear and I can expand.
> 
> OK, let me see how much I can refactor into the Memory class then we can see what to do about this. I don't completely understand the purpose of merging Image and Tensor and having an enum to differentiate. You would just end up replacing dynamic_cast with a test of the enum, which I imagine is what the compiler ends up doing internally to do the dynamic_cast.

Just to clarify, it wouldn't be merging the entire image and tensor class, but only the imageDataType and tensorDataType into a single enum `DataType` - although I assume that's what you meant. The main reason I was suggesting this is to allow for the Memory class to have a top level virutal `Memory::DataType datatype()` funcition that can then can be called directly through the parent class without having to convert, and hence being able to implement the respective operators. 

In regards to the dynamic_cast, I have to say that overall I don't have big issues with this implementation, as I liked how it was used - however seeing the need for a dynamic cast makes me think that having an explicit `memoryImplementation` variable that specifies whether it's an image vs a tensor could be valuable so the check can be done explicitly instead of requiring a dynamic_cast in case users don't need to completely cast it to perform an operation.

File: src/Image.cpp
Comment (robquill) [reply src/Image.cpp]: This isn't required and has been removed now that Tensors and Images use the same enum for their data types (Memory::DataType)

**Comment by robquill** [comment]
g

> > > > I'm not sure how to make this a compile-time error because the dtype is passed as a variable.
> > > 
> > > 
> > > Apologies, it seems I didn't share full context. In the current implementation there's nothing that would have to be done. However following the [suggestion here](https://github.com/KomputeProject/kompute/pull/388#discussion_r1720949108) of merging the imageType and tensorType into a top level enum it would be required as it would by default have an `eCustom` enum value (as this is possible for the Tensor class), so for the image it would be required to restrict it, which could be done by using something like [std::is_base_of](https://en.cppreference.com/w/cpp/types/is_base_of).
> > > However indeed it would only be necessary given the suggestion above - although hasn't yet been discussed. Let me know if this is not clear and I can expand.
> > 
> > 
> > OK, let me see how much I can refactor into the Memory class then we can see what to do about this. I don't completely understand the purpose of merging Image and Tensor and having an enum to differentiate. You would just end up replacing dynamic_cast with a test of the enum, which I imagine is what the compiler ends up doing internally to do the dynamic_cast.
> 
> Just to clarify, it wouldn't be merging the entire image and tensor class, but only the imageDataType and tensorDataType into a single enum `DataType` - although I assume that's what you meant. The main reason I was suggesting this is to allow for the Memory class to have a top level virutal `Memory::DataType datatype()` funcition that can then can be called directly through the parent class without having to convert, and hence being able to implement the respective operators.
> 
> In regards to the dynamic_cast, I have to say that overall I don't have big issues with this implementation, as I liked how it was used - however seeing the need for a dynamic cast makes me think that having an explicit `memoryImplementation` variable that specifies whether it's an image vs a tensor could be valuable so the check can be done explicitly instead of requiring a dynamic_cast in case users don't need to completely cast it to perform an operation.

OK, the latest code has imageDataType and tensorDataType replaced with a single enum Memory::DataType. As part of that I added support for Tensors of u/int8/16_t as these formats were already supported for images.

File: src/include/kompute/Image.hpp
Comment (axsaucedo) [reply src/include/kompute/Image.hpp:46]: Understood

File: src/Image.cpp
Comment (axsaucedo) [reply src/Image.cpp:313]: Ok makes sense

File: src/include/kompute/operations/OpImageSyncLocal.hpp
Comment (robquill) [reply src/include/kompute/operations/OpImageSyncLocal.hpp:63]: Done!

File: src/Image.cpp
Comment (robquill) [reply src/Image.cpp]: elementTypeSize has gone now, so resolving this one.

File: src/OpImageCopy.cpp
Comment (robquill) [reply src/OpImageCopy.cpp]: I've consolidated OpTensor/ImageCopy* into OpCopy in the latest revision.

File: src/include/kompute/Manager.hpp
Comment (robquill) [reply src/include/kompute/Manager.hpp]: You are right, this is not required. I've fixed it in the latest version.

File: src/Tensor.cpp
Comment (robquill) [reply src/Tensor.cpp]: I've removed this one as it's done in Memory.cpp.

File: src/Image.cpp
Comment (robquill) [reply src/Image.cpp]: I've implemented this test.

**Comment by robquill** [comment]
@axsaucedo , most of the major refactoring/consolidation is now done if you want to give this another review. There are some small bits like x,y still to do but it should be in a lot better shape now.

So far I don't think we need an enum for tensor/image but see  what you think and let me know what you want to do (if anything) with custom dtypes for images.

File: src/Image.cpp
Comment (robquill) [reply src/Image.cpp]: This was already done. I've removed the comment.

File: test/TestImage.cpp
Comment (axsaucedo) [reply test/TestImage.cpp:41]: There is the `__APPLE__` macro which could be used to easily skip

Comment (robquill) [reply test/TestImage.cpp:41]: I've dropped support for 3-channel images in the latest version as they are also not supported via the Mesa implementation either.

File: src/Image.cpp
Comment (robquill) [reply src/Image.cpp]: I've updated the code around this but there is still an underlying problem. When you create an image it's ImageLayout is eUndefined. Before you use it (including creating descriptor sets using it) you should transition it to eGeneral. However, to do this you need to do a memory barrier and a memory barrier needs a command buffer, which we don't have at that time.

File: src/Algorithm.cpp
Comment (robquill) [reply src/Algorithm.cpp]: I've fixed this in the latest version.

**Review by axsaucedo — CHANGES_REQUESTED** [review]
This looks good, overall quite clean as well without major modifications to the API, as it seems most examples work as is (only with the OpTensorX -> OpX change). I added a few comments but these are minor. We should be able to move to finalising the last few admin points (docs, codecov, etc) and wrapping up in order to merge. Thank you for the contribution!

File: src/include/kompute/Memory.hpp
Comment (robquill) [reply src/include/kompute/Memory.hpp]: Fixed

File: src/Memory.cpp
Comment (robquill) [reply src/Memory.cpp]: Done. I ended up following the method from here: https://stackoverflow.com/questions/22288347/dynamic-cast-vs-virtual-function-vs-member-enum. I didn't realise dynamic cast was slower, I thought the compiler did the equivalent of having an enum under the hood, but I guess it's more complicated in a more complex class hierarchy or something like that which slows it down.

File: src/include/kompute/Memory.hpp
Comment (robquill) [reply src/include/kompute/Memory.hpp:27]: Done using the method from here: https://stackoverflow.com/questions/22288347/dynamic-cast-vs-virtual-function-vs-member-enum

File: src/Image.cpp
Comment (robquill) [reply src/Image.cpp:29]: Done

**Comment by robquill** [comment]
Code coverage generally looks good (see attached). Most of the lines that aren't reached are because they are error cases that are difficult/impossible to hit. See the attached report.
[html.zip](https://github.com/user-attachments/files/16836958/html.zip)

File: python/test/test_image_types.py
Comment (robquill) [reply python/test/test_image_types.py:157]: I was intrigued by this too. According to pdb "params" is a List of kp.Image objects:

(Pdb) p params
[<kp.Image object at 0x7f229b4d8c30>, <kp.Image object at 0x7f229b4d9030>, <kp.Image object at 0x7f22addc1b30>]

and according to https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html a List can be automatically converted to std::vector. The vector type will be kp::Memory as this is what the Op* functions take, so I imagine that each kp::Image is cast to a kp::Memory when it is copied from the List to the std::vector.

**Comment by robquill** [comment]
@axsaucedo , does this updated diagram look OK to you? https://drive.google.com/file/d/1iKMY9YPKT97WgjYSYQ8LNHmRyXPwwx-H/view?usp=sharing

File: src/Memory.cpp
Comment (axsaucedo) [reply src/Memory.cpp]: Interesting - this is a great resource, I tried to find something but didn't find an answer as clear as this.

File: python/test/test_image_types.py
Comment (axsaucedo) [reply python/test/test_image_types.py:157]: Ok this is quite interesting and good to know, also being an implicit conversion

File: src/include/kompute/Image.hpp
Comment (axsaucedo) [inline src/include/kompute/Image.hpp:118]: Is this check necessary given the static_assert below?

**Review by axsaucedo — APPROVED** [review]
LGTM! We can update documentation as a separate PR + with respective naming convention - DCO check is missing so we can merge.

**Comment by robquill** [comment]
@axsaucedo , I've made a bit of a mess of the rebase by trying to fix the DCO. The end state should be unchanged but I'm thinking of just squashing all the commits into one as now the middle is a bit of a mess. Any thoughts/objections before I go ahead and do it?

**Pull request merged**
